File: src/menu/menues/ClientGameInterface.java
package menu.menues;

import java.net.InetAddress;
import java.util.LinkedList;

import core.Main;
import game.ClientGame;
import network.game.packets.GameFrameUpdatePacket;
import tilemap.LobbyTileMap;
import misc.Debug;
import network.Packet;
import network.game.packets.EventPacket;
import player.LobbyPlayer;

public class ClientGameInterface extends GameInterface
{
	private InetAddress serverIP;

	public ClientGameInterface(LobbyTileMap map, LinkedList<LobbyPlayer> lobbyPlayers, int localPlayerID, InetAddress serverIP)
	{
		super(new ClientGame(map, lobbyPlayers, localPlayerID));
		Debug.warnIf(serverIP == null, "ClientGameInterface.<init>(): serverIP is null");
		this.serverIP = serverIP;
	}

	private ClientGame getGame() { return (ClientGame) getUncastedGame(); }

	@Override public void handlePacket(Packet packet, InetAddress ip)
	{
		if (packet instanceof GameFrameUpdatePacket)
		{
			getGame().apply((GameFrameUpdatePacket) packet);
		}
		else
		{
			Debug.warn("ClientGameInterface.handlePacket(): Unknown Packet: " + packet.toString());
		}
	}

	@Override public void render()
	{
		super.render();
	}

	@Override public void onEvent(EventPacket packet)
	{
		super.onEvent(packet);
		Main.getNetworkDevice().send(packet, serverIP);
	}
}

File: cpp/src/game/ClientGameInterface.cpp
#include "ClientGameInterface.hpp"

#include <misc/Global.hpp>
#include <collision/CollisionTester.hpp>
#include <player/GamePlayer.hpp>
#include <entity/Idler.hpp>
#include <entity/Tile.hpp>
#include <tilemap/GameTileMap.hpp>
#include <math/game/GameRect.hpp>
#include <network/packets/ChangeActionsPacket.hpp>
#include <network/packets/PacketWithID.hpp>

ClientGameInterface::ClientGameInterface(LobbyTileMap* map, const std::vector<LobbyPlayer*>& players, int playerID, const sf::IpAddress& ip, long int unixTime_arg)
	: GameInterface(map, players, unixTime_arg), localPlayerID(playerID)
{
	serverIP = ip;
}

ClientGameInterface::~ClientGameInterface()
{}

void ClientGameInterface::handlePacket(Packet* packet, const sf::IpAddress& ipAddress)
{
	if (ipAddress != serverIP) {
		Debug::warn("received packet, not coming from server");
		delete packet;
		return;
	}

	switch (packet->getCompressID())
	{
		case CompressIDs::PACKET_WITH_ID:
		{
			PacketWithID* pwid = packet->unwrap<PacketWithID>();
			switch (pwid->getPacket()->getCompressID())
			{
				case CompressIDs::CHANGE_ACTIONS_PACKET:
				{
					ChangeActionsPacket* cap = pwid->getPacket()->unwrap<ChangeActionsPacket>();
					addCalendarEntry(cap->getFrameNumber(), pwid->getID(), cap->getActions());
					backtrack();
					delete cap;
					break;
				}
				default:
				{
					Debug::error("ClientGameInterface::handlePacket(): unknown Packet (in PacketWithID) with CompressID=" + Converter::intToString((int) pwid->getPacket()->getCompressID()));
				}
			}
			break;
		}
		default:
		{
			Debug::error("ClientGameInterface::handlePacket(): unknown Packet with CompressID=" + Converter::intToString((int) packet->getCompressID()));
		}
	}
	delete packet;
}

void ClientGameInterface::subTick()
{
	handleAllPackets();

	Actions a = calcActions();
	if (serverActionsStatus != a)
	{
		// Send
		serverActionsStatus = a;
		ChangeActionsPacket packet(a, getFrameCounter());
		send(&packet, serverIP);
		// add to Calendar
		addCalendarEntry(getFrameCounter(), localPlayerID, a);
	}

	applyCalendar();

	mainFrame->tick();
}

GamePlayer* ClientGameInterface::getLocalPlayer() const
{
	return mainFrame->players[localPlayerID];
}
